source/val/validate_invalid_type.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2025 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Performs validation of invalid type instructions.
+
+#include <vector>
+
+#include "source/opcode.h"
+#include "source/val/instruction.h"
+#include "source/val/validate.h"
+#include "source/val/validation_state.h"
+
+namespace spvtools {
+namespace val {
+
+// Validates correctness of certain special type instructions.
+spv_result_t InvalidTypePass(ValidationState_t& _, const Instruction* inst) {
```

Comment:
This still feels incomplete. For example, `OpExtInst` is not rejected.

Maybe a better approach would be:
```
if (spvOpcodeGeneratesType(inst->opcode()) {
  return SPV_SUCCESS;
}

switch (inst->opcode()) {
  // allow listed instructions
  case OpLoad:
  // ...
  case OpStore:
    return SPV_SUCCESS;
  default:
    break;
}

for (uint32_t i = 0; i < inst->operands().size(); i++) {
  const auto  &op = inst->operand(i);
  const auto &op_type = op.type;
  if (spvIsIdType(op_type)) {
    const auto op_type_id = _.GetOperandTypeId(inst, i);
    if (_.IsBfloat16ScalarType(op_type_id) ||
        _.IsBfloat16VectorType(op_type_id)) {
      return _.diag(SPV_ERROR_INVALID_DATA, inst) << spvOpcodeString(inst->opcode()) << " doesn't support Bfloat16";
    }
  }
}
```

Reply:
That's a good idea!

Reply:
I pulled this into my glslang change and it found that at least the following cases are missing:

```
    case spv::Op::OpName:
    case spv::Op::OpDecorate:
    case spv::Op::OpFunctionCall:
    case spv::Op::OpFunction:
    case spv::Op::OpBitcast:
```

Reply:
Fixed, thank you.

Reply:
Have you tried running this against CTS? I think there are a lot more missing here, including most of the cooperative matrix and cooperative matrix 2 opcodes and the SPV_EXT_replicated_composites instructions.

Reply:
Indeed, it's very easy to miss some `spv::OP` here. I'll test them later. Thank you.

Reply:
https://github.com/KhronosGroup/SPIRV-Registry/pull/323 misses annotation (OpDecorate) and debug (OpName) instructions in the list of valid uses. For that matter whether miscellaneous (OpUndef) and extension (OpExtInst future looking).

Reply:
I think I'll still change it to the reject lists. Just as @jeffbolznv said, there are still many extensions that need to support Bfloat16, but it's really difficult to list all the supported operations. Once one is missed, it may break the running of the program.

Reply:
We've filed #6067 to auto-generate predicates from the grammar. So for now feel free to add make this a deny list, but add a todo referencing that issue. I'll fix it up once we get the code in place.